src/main/scala/workflow/Pipeline.scala

```diff
@@ -0,0 +1,242 @@
+package workflow
+
+import org.apache.spark.rdd.RDD
+import pipelines.Logging
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+trait Pipeline[A, B] {
+  private[workflow] val nodes: Seq[Node]
+  private[workflow] val dataDeps: Seq[Seq[Int]]
+  private[workflow] val fitDeps: Seq[Seq[Int]]
+  private[workflow] val sink: Int
+
+  /** validates (returns an exception if false) that
+    - nodes.size = dataDeps.size == fitDeps.size
+
+    - there is a valid sink
+    - data nodes must have no deps
+    - estimators may not have fit deps, must have data deps
+    - transformers must have data deps, allowed to have fit deps
+
+    - data deps may not point at estimators
+    - fit deps may only point to estimators
+    */
+  private[workflow] def validate(): Unit = {
+    require(nodes.size == dataDeps.size && nodes.size == fitDeps.size, "nodes.size must equal dataDeps.size and fitDeps.size")
+    require(sink == Pipeline.SOURCE || (sink >= 0 && sink < nodes.size), "Sink must point at a valid node")
+
+    val nodeTuples = nodes.zip(dataDeps).zip(fitDeps).map(x => (x._1._1, x._1._2, x._2))
+    require(nodeTuples.forall(x => if (x._1.isInstanceOf[DataNode] && (x._2.nonEmpty || x._3.nonEmpty)) false else true), "DataNodes may not have dependencies")
```

Comment:
Yeah, I'm also not sure about leaving this method in general. It could be useful, but I couldn't make it automatically run at construction time due to super constructor ordering stuff. Maybe I should move it to `ConcretePipeline`?